workspaces/ui-v2/src/components/sharing/ShareModal.tsx

```diff
@@ -0,0 +1,223 @@
+    // Ran it through https://jakearchibald.github.io/svgomg/
+    // base64 encoded it (on macos, you can do: `pbpaste | base64 | pbcopy` to quickly base64 encode whatever is in your paste buffer)
+    // prepend with `data:image/svg+xml;base64,`
+    logo:
+      'data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTM0IiBoZWlnaHQ9IjEzNCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB4bWxuczp4bGluaz0iaHR0cDovL3d3dy53My5vcmcvMTk5OS94bGluayI+PGRlZnM+PGxpbmVhckdyYWRpZW50IHgxPSI1MCUiIHkxPSItMTQuMyUiIHgyPSI1MCUiIHkyPSIxMDAlIiBpZD0iYyI+PHN0b3Agc3RvcC1jb2xvcj0iI0ZGRiIgb2Zmc2V0PSIwJSIvPjxzdG9wIHN0b3AtY29sb3I9IiNGNEY0RjQiIG9mZnNldD0iMTAwJSIvPjwvbGluZWFyR3JhZGllbnQ+PGZpbHRlciB4PSItNS42JSIgeT0iLTQlIiB3aWR0aD0iMTExLjElIiBoZWlnaHQ9IjExMS4xJSIgZmlsdGVyVW5pdHM9Im9iamVjdEJvdW5kaW5nQm94IiBpZD0iYSI+PGZlT2Zmc2V0IGR5PSIyIiBpbj0iU291cmNlQWxwaGEiIHJlc3VsdD0ic2hhZG93T2Zmc2V0T3V0ZXIxIi8+PGZlR2F1c3NpYW5CbHVyIHN0ZERldmlhdGlvbj0iMiIgaW49InNoYWRvd09mZnNldE91dGVyMSIgcmVzdWx0PSJzaGFkb3dCbHVyT3V0ZXIxIi8+PGZlQ29sb3JNYXRyaXggdmFsdWVzPSIwIDAgMCAwIDAgMCAwIDAgMCAwIDAgMCAwIDAgMCAwIDAgMCAwLjUgMCIgaW49InNoYWRvd0JsdXJPdXRlcjEiLz48L2ZpbHRlcj48Y2lyY2xlIGlkPSJiIiBjeD0iNjMiIGN5PSI2MyIgcj0iNjMiLz48L2RlZnM+PGcgZmlsbD0ibm9uZSIgZmlsbC1ydWxlPSJldmVub2RkIj48ZyBmaWxsLXJ1bGU9Im5vbnplcm8iIHRyYW5zZm9ybT0idHJhbnNsYXRlKDQgMikiPjx1c2UgZmlsbD0iIzAwMCIgZmlsdGVyPSJ1cmwoI2EpIiB4bGluazpocmVmPSIjYiIvPjx1c2UgZmlsbD0idXJsKCNjKSIgeGxpbms6aHJlZj0iI2IiLz48L2c+PHBhdGggZD0iTTkwIDM1YTM4IDM4IDAgMDAtNDYgMG0yOCA2OGEzOCAzOCAwIDAwMzMtMzhjMC04LTMtMTYtNy0yMm0tMzYgNjBhMzggMzggMCAwMS0zMy0zOGMwLTggMy0xNiA3LTIyIiBzdHJva2U9IiMzNDNCRTUiIHN0cm9rZS13aWR0aD0iNC43IiBzdHJva2UtbGluZWNhcD0icm91bmQiLz48cGF0aCBkPSJNNTEgODVhMjYgMjYgMCAwMDMyIDBNNjMgMzljLTUgMS0xMCAzLTE0IDdhMjYgMjYgMCAwMC0zIDM0bTI1LTQxYzUgMSAxMCAzIDE0IDdhMjYgMjYgMCAwMTMgMzQiIHN0cm9rZT0iIzM0M0JFNSIgc3Ryb2tlLXdpZHRoPSI0LjciIHN0cm9rZS1saW5lY2FwPSJyb3VuZCIvPjwvZz48L3N2Zz4=',
+  };
+
+  const queryParams = Object.entries(params)
+    .map(([k, v]) => `${k}=${encodeURIComponent(v)}`)
+    .join('&');
+
+  return `https://img.shields.io/static/v1?${queryParams}`;
+}
+
+function generateBadgeMarkdown({
+  endpointCount,
+  shareUrl,
+}: {
+  endpointCount: number;
+  shareUrl: string;
+}) {
+  return `[![API Docs: ${endpointCount} Endpoints](${generateBadgeUrl({
+    endpointCount,
+  })})](${shareUrl})`;
+}
+
+export const ShareModal: React.FC<{
+  share: (intent: ShareTarget) => Promise<string>;
+}> = (props) => {
+  const styles = useStyles(props);
+
+  const { isAuthenticated, isLoading, loginWithPopup } = useAuth0();
+
+  const endpointCount = useAppSelector(
+    (state) => state.endpoints.results?.data!.length || 0
+  );
+
+  const [shareType, setShareType] = useState<ShareTarget | null>(null);
+
+  let inputRef = useRef<HTMLInputElement>();
+  const [copied, setCopied] = useState(false);
+
+  useEffect(() => {
+    if (copied) {
+      let timeout = setTimeout(() => setCopied(false), 7500);
+      return () => clearTimeout(timeout);
+    }
+  }, [copied, setCopied]);
```

Comment:
So this works, but it's probably not ideal - what might be better here is to add this setTimeout to the end of the `copy` function (otherwise you have to handle things like `if copied` and `return () => clearTimeout` which can be easy to miss things and cause infinite loops!

Reply:
Happy to move it there, like this?

```typescript
  const copy = useCallback(() => {
    const input = inputRef.current;
    if (input) {
      input.focus();
      input.select();
      document.execCommand('copy');
      setCopied(true);
      setTimeout(() => setCopied(false), 7500);
    }
  }, []);
```

I do wonder though, don't I still need to `return () => clearTimeout` here?

Reply:
Bleh, ok I think what you have is fine - I was trying to avoid changed useEffects - but in this case if we have a setTimeout with setState, if this unmounts we have a memoryleak (it's likely not a big impact, apart from having a react warning.

This is probably _fine_ to keep as is

Reply:
What's the memory leak? `return () => clearTimeout(timeout);` cancells the `setTimeout` on unmount, righT?